README.md
# LightCrop

Lightweight image editor built with Python and FFmpeg: aiming for fast app launch, minimal dependencies, and a straightforward interface.

![banner](banner.jpg)


## Installation

- Manual
    1. Download the latest stable version [here](https://github.com/nvfp/LightCrop/releases)
    2. Extract and put it in your favorite folder (e.g., `~/myprojects/LightCrop-x.x.x`)
    3. (optional) Remove the version number (`~/myprojects/LightCrop-x.x.x` -> `~/myprojects/LightCrop`)
    3. Install dependencies:
        - [carbon](https://github.com/nvfp/carbon) (1.2.0 or higher)
        - [FFmpeg](https://ffmpeg.org/download.html)
    4. Open terminal, navigate to `~/myprojects` folder, and run `python LightCrop`


## Usage

- Run:    

    ```sh
    python LightCrop
    ```

- Open configuration file:

    ```sh
    python LightCrop settings
    ```

- To exit the software: press `esc` key


## Troubleshoot

- To report bugs or ask questions, you can open an issue or submit a pull request.


## License

This project is licensed under the MIT license.


_pages/Changelog.md
---
permalink: /changelog/
layout: main
title: Changelog
---

# Changelog



main/core.py
import datetime
import math
import os
import subprocess as sp
import tkinter as tk
from PIL import ImageTk, Image
from tkinter import filedialog

from carbon.gui.button.v2 import Button
from carbon.gui.slider import Slider
from carbon.utils import printer

from main.constants import ALLOWED_EXTENSIONS, PROXY_FILE_PTH


def render(
    ffmpeg,
    input_pth,
    output_pth,

    rotate,

    do_crop,
    crop_w,
    crop_h,
    crop_x,
    crop_y,

    do_contrast,
    do_brightness,
    do_saturation,
    do_gamma,
    do_gamma_r,
    do_gamma_g,
    do_gamma_b,
    do_vignette,
    do_colortemperature,
    do_avgblur,

    contrast,
    brightness,
    saturation,
    gamma,
    gamma_r,
    gamma_g,
    gamma_b,
    vignette,
    colortemperature,
    avgblur,

    q_v
):

    filters = []

    if rotate == 1:  # 90 CCW
        filters.append('transpose=2')
    elif rotate == 2:  # 180 deg
        filters.append('transpose=1,transpose=1')
    elif rotate == 3:  # 90 CW
        filters.append('transpose=1')

    if do_crop:
        filters.append(f'crop={crop_w}:{crop_h}:{crop_x}:{crop_y}')

    if do_contrast or do_brightness or do_saturation or do_gamma or do_gamma_r or do_gamma_g or do_gamma_b:
        eqs = []
        if do_contrast:
            eqs.append(f'contrast={contrast}')
        if do_brightness:
            eqs.append(f'brightness={brightness}')
        if do_saturation:
            eqs.append(f'saturation={saturation}')
        if do_gamma:
            eqs.append(f'gamma={gamma}')
        if do_gamma_r:
            eqs.append(f'gamma_r={gamma_r}')
        if do_gamma_g:
            eqs.append(f'gamma_g={gamma_g}')
        if do_gamma_b:
            eqs.append(f'gamma_b={gamma_b}')
        filters.append('eq=' + ':'.join(eqs))

    if do_vignette:
        if vignette >= 0:
            filters.append(f'vignette=a={math.pi*vignette/180}:mode=backward')
        else:
            filters.append(f'vignette=a={-math.pi*vignette/180}:mode=forward')
    
    if do_colortemperature:
        filters.append(f'colortemperature=temperature={colortemperature}')
    
    if do_avgblur:
        filters.append(f'avgblur={avgblur}')


    if filters == []:
        filters_cmd = []
    else:
        filters_cmd = ['-vf', ','.join(filters),]

    cmd = [
        ffmpeg, '-v', 'error',
        '-i', input_pth,
        *filters_cmd,
        '-q:v', str(q_v),
        '-y',  # overwrite
        output_pth
    ]
    printer(f'INFO: Rendering {repr(output_pth)}...')
    sp.call(cmd)


def reshow_proxy_photo(SETTINGS_FFMPEG, page: tk.Canvas, Rt, PROXY_BOX_W, PROXY_BOX_H, PROXY_BOX_X, PROXY_BOX_Y):

    render(
        SETTINGS_FFMPEG,
        Rt.input_pth,
        PROXY_FILE_PTH,

        Rt.rotate,

        False,
        Rt.crop_w,
        Rt.crop_h,
        Rt.crop_x,
        Rt.crop_y,

        Rt.do_contrast,
        Rt.do_brightness,
        Rt.do_saturation,
        Rt.do_gamma,
        Rt.do_gamma_r,
        Rt.do_gamma_g,
        Rt.do_gamma_b,
        Rt.do_vignette,
        Rt.do_colortemperature,
        Rt.do_avgblur,

        Rt.contrast,
        Rt.brightness,
        Rt.saturation,
        Rt.gamma,
        Rt.gamma_r,
        Rt.gamma_g,
        Rt.gamma_b,
        Rt.vignette,
        Rt.colortemperature,
        Rt.avgblur,

        Rt.q_v
    )
    
    img = Image.open(PROXY_FILE_PTH)
    w = img.width
    h = img.height
    r = w/h

    if r > (PROXY_BOX_W/PROXY_BOX_H):  # landscape
        W = round(PROXY_BOX_W)
        H = round(W/r)
        X = PROXY_BOX_X
        Y = PROXY_BOX_Y + (PROXY_BOX_H - H)/2
    else:  # portrait
        H = round(PROXY_BOX_H)
        W = round(H*r)
        X = PROXY_BOX_X + (PROXY_BOX_W - W)/2
        Y = PROXY_BOX_Y

    ## rescale
    img = img.resize((W, H))
    Rt.proxy_photo = ImageTk.PhotoImage(img)  # to prevent the image from being collected by the Python garbage collector

    page.delete('proxy_photo')
    page.create_image(X, Y, image=Rt.proxy_photo, anchor='nw', tags='proxy_photo')
    page.create_rectangle(X-1, Y-1, X+W, Y+H, outline='#555', tags='proxy_photo')  # border

    ## cropping needs
    Rt.crop_tl_x = X
    Rt.crop_tl_y = Y
    Rt.crop_dr_x = X + W
    Rt.crop_dr_y = Y + H
    Rt.crop_proxy_w = W
    Rt.crop_proxy_h = H
    Rt.crop_proxy_x = X
    Rt.crop_proxy_y = Y
    Rt.crop_real_w = w
    Rt.crop_real_h = h


def core(
    page: tk.Canvas,
    Rt, SETTINGS_FFMPEG, SETTINGS_OPEN_DIR_PTH, SETTINGS_SAVE_DIR_PTH,
    PROXY_BOX_W, PROXY_BOX_H, PROXY_BOX_X, PROXY_BOX_Y,
    redraw_crop_grid
):

    def rotate_left():
        Rt.rotate = (Rt.rotate + 1) % 4
        reshow_proxy_photo(SETTINGS_FFMPEG, page, Rt, PROXY_BOX_W, PROXY_BOX_H, PROXY_BOX_X, PROXY_BOX_Y)
    
    def rotate_right():
        Rt.rotate = (Rt.rotate - 1) % 4
        reshow_proxy_photo(SETTINGS_FFMPEG, page, Rt, PROXY_BOX_W, PROXY_BOX_H, PROXY_BOX_X, PROXY_BOX_Y)

    Button(
        x=50, y=55,
        fn=rotate_left, label='Rotate (Left)', id='rotate_ccw', width=85, locked=True, tags='tools'
    )
    Button(
        x=Button.get_bounding_box_by_id('rotate_ccw')[2]+15,
        y=Button.get_bounding_box_by_id('rotate_ccw')[1],
        fn=rotate_right, label='Rotate (Right)', id='rotate_cw', width=85, locked=True, tags='tools'
    )


    def crop_toggle():

        Rt.do_crop = not Rt.do_crop

        if Rt.do_crop:
            Button.set_label_by_id('crop', 'Crop (ON)')
            # Button.set_lock_by_tag('crop_ratio', False)
            redraw_crop_grid()
        else:
            Button.set_label_by_id('crop', 'Crop (OFF)')
            # Button.set_lock_by_tag('crop_ratio', True)
            page.delete('crop_grid')

    Button(
        x=50,
        y=Button.get_bounding_box_by_id('rotate_cw')[3]+25,
        fn=crop_toggle, label='Crop (OFF)', id='crop', width=70, locked=True, tags='tools'
    )
    Button(
        x=Button.get_bounding_box_by_id('crop')[2]+10,
        y=Button.get_bounding_box_by_id('crop')[1],
        fn=None, label='1:1', id='crop1:1', width=30, locked=True, tags='crop_ratio'
    )
    Button(
        x=Button.get_bounding_box_by_id('crop1:1')[2]+5,
        y=Button.get_bounding_box_by_id('crop')[1],
        fn=None, label='16:9', id='crop16:9', width=30, locked=True, tags='crop_ratio'
    )
    Button(
        x=Button.get_bounding_box_by_id('crop16:9')[2]+5,
        y=Button.get_bounding_box_by_id('crop')[1],
        fn=None, label='9:16', id='crop9:16', width=30, locked=True, tags='crop_ratio'
    )
    Button(
        x=Button.get_bounding_box_by_id('crop9:16')[2]+5,
        y=Button.get_bounding_box_by_id('crop')[1],
        fn=None, label='2:1', id='crop2:1', width=30, locked=True, tags='crop_ratio'
    )
    Button(
        x=Button.get_bounding_box_by_id('crop2:1')[2]+5,
        y=Button.get_bounding_box_by_id('crop')[1],
        fn=None, label='1:2', id='crop1:2', width=30, locked=True, tags='crop_ratio'
    )


    X = 20
    Y = Button.get_bounding_box_by_id('crop')[3] + 65
    GAP = 45
    GAP2 = 120  # gap between gate and slider

    def update_filter_gate(filter_name):
        
        filter_gate = 'do_' + filter_name
        setattr(Rt, filter_gate, not getattr(Rt, filter_gate))  # flipping the value
        
        if getattr(Rt, filter_gate):
            Button.set_label_by_id(filter_gate, 'Used')
            Slider.set_lock_by_id(filter_name, False)
        else:
            Button.set_label_by_id(filter_gate, 'OFF')
            Slider.set_lock_by_id(filter_name, True)
        
        reshow_proxy_photo(SETTINGS_FFMPEG, page, Rt, PROXY_BOX_W, PROXY_BOX_H, PROXY_BOX_X, PROXY_BOX_Y)

    def update_filter_value(filter_name):
        new_value = Slider.get_value_by_id(filter_name)
        setattr(Rt, filter_name, new_value)

        reshow_proxy_photo(SETTINGS_FFMPEG, page, Rt, PROXY_BOX_W, PROXY_BOX_H, PROXY_BOX_X, PROXY_BOX_Y)

    BTN_W = 40  # gate buttons width
    
    Button(x=X, y=Y+GAP*0, fn=lambda: update_filter_gate('contrast'), label='OFF', id='do_contrast', width=BTN_W, locked=True, anchor='w', tags='tools')
    Slider(
        id='contrast',
        min=-5, max=5, step=0.01, init=Rt.contrast,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*0,
        fn=lambda: update_filter_value('contrast'),
        label='Contrast',
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, locked=True
    )
    Button(x=X, y=Y+GAP*1, fn=lambda: update_filter_gate('brightness'), label='OFF', id='do_brightness', width=BTN_W, locked=True, anchor='w', tags='tools')
    Slider(
        id='brightness',
        min=-1, max=1, step=0.005, init=Rt.brightness,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*1,
        fn=lambda: update_filter_value('brightness'),
        label='Brightness',
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, locked=True
    )
    Button(x=X, y=Y+GAP*2, fn=lambda: update_filter_gate('saturation'), label='OFF', id='do_saturation', width=BTN_W, locked=True, anchor='w', tags='tools')
    Slider(
        id='saturation',
        min=0, max=3, step=0.01, init=Rt.saturation,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*2,
        fn=lambda: update_filter_value('saturation'),
        label='Saturation',
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, locked=True
    )
    Button(x=X, y=Y+GAP*3, fn=lambda: update_filter_gate('gamma'), label='OFF', id='do_gamma', width=BTN_W, locked=True, anchor='w', tags='tools')
    Slider(
        id='gamma',
        min=0.1, max=10, step=0.005, init=Rt.gamma,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*3,
        fn=lambda: update_filter_value('gamma'),
        label='Gamma',
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, locked=True
    )
    Button(x=X, y=Y+GAP*4, fn=lambda: update_filter_gate('gamma_r'), label='OFF', id='do_gamma_r', width=BTN_W, locked=True, anchor='w', tags='tools')
    Slider(
        id='gamma_r',
        min=0.1, max=10, step=0.005, init=Rt.gamma_r,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*4,
        fn=lambda: update_filter_value('gamma_r'),
        label='Gamma-R',
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, locked=True
    )
    Button(x=X, y=Y+GAP*5, fn=lambda: update_filter_gate('gamma_g'), label='OFF', id='do_gamma_g', width=BTN_W, locked=True, anchor='w', tags='tools')
    Slider(
        id='gamma_g',
        min=0.1, max=10, step=0.005, init=Rt.gamma_g,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*5,
        fn=lambda: update_filter_value('gamma_g'),
        label='Gamma-G',
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, locked=True
    )
    Button(x=X, y=Y+GAP*6, fn=lambda: update_filter_gate('gamma_b'), label='OFF', id='do_gamma_b', width=BTN_W, locked=True, anchor='w', tags='tools')
    Slider(
        id='gamma_b',
        min=0.1, max=10, step=0.005, init=Rt.gamma_b,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*6,
        fn=lambda: update_filter_value('gamma_b'),
        label='Gamma-B',
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, locked=True
    )
    Button(x=X, y=Y+GAP*7, fn=lambda: update_filter_gate('vignette'), label='OFF', id='do_vignette', width=BTN_W, locked=True, anchor='w', tags='tools')
    Slider(
        id='vignette',
        min=-90, max=90, step=1, init=Rt.vignette,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*7,
        fn=lambda: update_filter_value('vignette'),
        label='Vignette',
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, locked=True
    )
    Button(x=X, y=Y+GAP*8, fn=lambda: update_filter_gate('colortemperature'), label='OFF', id='do_colortemperature', width=BTN_W, locked=True, anchor='w', tags='tools')
    Slider(
        id='colortemperature',
        min=1000,
        # max=40000,  # is the maximum that ffmpeg allows
        max=20000,  # should be enough
        step=50, init=Rt.colortemperature,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*8,
        fn=lambda: update_filter_value('colortemperature'),
        label='White Balance',
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, suffix=' K', locked=True
    )
    Button(x=X, y=Y+GAP*9, fn=lambda: update_filter_gate('avgblur'), label='OFF', id='do_avgblur', width=BTN_W, locked=True, anchor='w', tags='tools')
    Slider(
        id='avgblur',
        min=1,
        # max=1024,  # is the maximum that ffmpeg allows
        max=25,  # should be enough
        step=1, init=Rt.avgblur,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*9,
        fn=lambda: update_filter_value('avgblur'),
        label='Blur',  # average blur
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, locked=True
    )
    Slider(
        id='q_v',
        min=1, max=31, step=1, init=Rt.q_v,
        x=Button.get_bounding_box_by_id('do_contrast')[2]+GAP2, y=Y+GAP*10,
        fn=lambda: update_filter_value('q_v'),
        label='Quality (1: highest, 31: lowest)',
        label_y_shift=-20, yield_x_shift=40, yield_box_color='#000', yield_box_height=20, yield_box_width=55, show_yield_box=True, locked=True, tags='tools'
    )


    def open_new_image():

        pth = filedialog.askopenfilename(initialdir=SETTINGS_OPEN_DIR_PTH)
        if (not os.path.isfile(pth)) or (not pth.lower().endswith(ALLOWED_EXTENSIONS)):
            printer(f'WARNING: Invalid image: {repr(pth)}')
            return

        Rt.input_pth = pth

        Button.set_lock_by_tag('tools', False)
        Slider.set_lock_by_tag('tools', False)
        
        reshow_proxy_photo(SETTINGS_FFMPEG, page, Rt, PROXY_BOX_W, PROXY_BOX_H, PROXY_BOX_X, PROXY_BOX_Y)

    Button(
        x=50, y=Y+GAP*11+25,
        fn=open_new_image, label='Open', id='open'
    )

    def save_the_output():

        name, ext = os.path.splitext(os.path.basename(Rt.input_pth))
        date = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')

        output_pth = filedialog.asksaveasfilename(
            initialfile=f'{name}_{date}{ext}',
            initialdir=SETTINGS_SAVE_DIR_PTH,
            filetypes=(
                ('JPEG', '*.jpg'),
                ('JPEG', '*.jpeg'),
                ('PNG', '*.png'),
                ('Bitmap', '*.bmp'),
                ('TIFF', '*.tiff'),
                ('WebP', '*.webp'),
                ('SVG', '*.svg'),
                ('ICO', '*.ico'),
                ('HEIC', '*.heic'),
            )
        )

        if os.path.exists(output_pth):
            printer(f'WARNING: Unable to render the output; the file already exists: {repr(output_pth)}')
            return
        
        render(
            SETTINGS_FFMPEG,
            Rt.input_pth,
            output_pth,

            Rt.rotate,

            Rt.do_crop,
            Rt.crop_w,
            Rt.crop_h,
            Rt.crop_x,
            Rt.crop_y,

            Rt.do_contrast,
            Rt.do_brightness,
            Rt.do_saturation,
            Rt.do_gamma,
            Rt.do_gamma_r,
            Rt.do_gamma_g,
            Rt.do_gamma_b,
            Rt.do_vignette,
            Rt.do_colortemperature,
            Rt.do_avgblur,

            Rt.contrast,
            Rt.brightness,
            Rt.saturation,
            Rt.gamma,
            Rt.gamma_r,
            Rt.gamma_g,
            Rt.gamma_b,
            Rt.vignette,
            Rt.colortemperature,
            Rt.avgblur,

            Rt.q_v
        )

        printer(f'INFO: Output is saved: {repr(output_pth)}')

    Button(
        x=Button.get_bounding_box_by_id('open')[2]+15,
        y=Button.get_bounding_box_by_id('open')[1],
        fn=save_the_output, anchor='nw', label='Save', locked=True, tags='tools'
    )

main/__init__.py
import os
import subprocess as sp
import sys
import tkinter as tk

from carbon.gui.button.v2 import Button
from carbon.gui.label import Label
from carbon.gui.slider import Slider
from carbon.keycrate import KeyCrate
from carbon.path import open_file
from carbon.utils import printer

from main.constants import SOFTWARE_NAME, SOFTWARE_VER, TMP_DIR_PTH, SETTINGS_FILE_PTH, PROXY_FILE_PTH
from main.core import core


def startup_checkup():

    ## When users force-exit (not via 'esc' key), the proxy file won't be deleted.
    ## Also, if there's an error or something unexpected happens,
    ## the proxy file might not be deleted either.
    ## So, the following code makes sure the "tmp" folder stays clean in such cases.
    if os.path.isfile(PROXY_FILE_PTH):
        printer(f'WARNING: Found proxy file from previous app usage. Deleting {repr(PROXY_FILE_PTH)}...')
        os.remove(PROXY_FILE_PTH)

    ## "tmp" folder should be clean, used for storing intermediate files
    ## during runtime and to prevent unintended file deletions.
    if not (
        (len(os.listdir(TMP_DIR_PTH)) == 1)
        and
        (os.listdir(TMP_DIR_PTH)[0] == '.gitkeep')
    ):
        ## this shouldn't be raised unless bug occurs
        raise AssertionError(f'Directory {repr(TMP_DIR_PTH)} is not clean.')

startup_checkup()


def parse_settings():

    ## if users attempt to open settings
    if len(sys.argv) != 1:
        if sys.argv[1] == 'settings':
            open_file(SETTINGS_FILE_PTH)
            printer(f'INFO: settings file opened')
        else:
            printer(f'ERROR: Maybe you meant to run `python {SOFTWARE_NAME} settings` instead?')
        sys.exit(1)


    ## parsing the settings
    try:
        settings = KeyCrate(
            SETTINGS_FILE_PTH, key_is_var=True, eval_value=True,
            only_keys=['ffmpeg', 'open_dir', 'save_dir']
        )
    except (SyntaxError, ValueError, AssertionError) as err:
        printer(f'ERROR: Run `python {SOFTWARE_NAME} settings` to fix this error: {err}')
        sys.exit(1)

    ## <checking ffmpeg>
    ffmpeg = settings.ffmpeg

    if ffmpeg != 'ffmpeg':
        if not (os.path.isfile(ffmpeg) and os.path.splitext(ffmpeg.lower())[1] == '.exe'):
            printer(f'ERROR: FFmpeg not recognized or not an .exe file (Run `python {SOFTWARE_NAME} settings` to fix): {repr(ffmpeg)}')
            sys.exit(1)

    try:
        sp.run([ffmpeg, '-version'], stdout=sp.DEVNULL, stderr=sp.DEVNULL)
        printer(f'INFO: ffmpeg is valid and usable.')
    except FileNotFoundError:
        printer(f'ERROR: ffmpeg not found or not a recognized command (Run `python {SOFTWARE_NAME} settings` to fix): {repr(ffmpeg)}')
        sys.exit(1)
    ## </checking ffmpeg>

    ## <validating open/save dir>
    open_dir = settings.open_dir
    if open_dir != '/':
        if not os.path.isdir(open_dir):
            printer(f'WARNING: the default open-folder is not a dir: {repr(open_dir)}')
            open_dir = '/'

    save_dir = settings.save_dir
    if save_dir != '/':
        if not os.path.isdir(save_dir):
            printer(f'WARNING: the default save-folder is not a dir: {repr(save_dir)}')
            save_dir = '/'
    ## </validating open/save dir>

    return ffmpeg, open_dir, save_dir

SETTINGS_FFMPEG, SETTINGS_OPEN_DIR_PTH, SETTINGS_SAVE_DIR_PTH = parse_settings()


root = tk.Tk()
root.title(f'{SOFTWARE_NAME}-v{SOFTWARE_VER}')
root.attributes('-fullscreen', True)

MON_W = root.winfo_screenwidth()
MON_H = root.winfo_screenheight()

page = tk.Canvas(width=MON_W, height=MON_H, bg='#111', highlightthickness=0, borderwidth=0)
page.place(x=0, y=0)

Button.set_page(page)
Slider.set_page(page)

Slider.set_page_focus([None])  # not using page-focus mechanism


def prepare():  # using a function to preserve variable names and avoid conflicts
    """will return the variable that is going to be used"""
    
    r = 0.28  # to adjust the tools page width
    page.create_rectangle(
        MON_W*r, -1,  # -1 instead of 0 to remove the top border
        MON_W, MON_H,
        fill='#070707', outline='#555'
    )
    
    padx = 0.05
    pady = 0.05

    w = (1 - padx*2)*(MON_W - MON_W*r)
    h = (1 - pady*2)*MON_H
    x = MON_W*r + ((MON_W - MON_W*r) - w)*0.5  # 0.5 to make it centered
    y = (MON_H - h)*0.5
    
    ## uncomment to show the image bounding box
    # page.create_rectangle(x, y, x+w, y+h, outline='#f00')

    return w, h, x, y

PROXY_BOX_W, PROXY_BOX_H, PROXY_BOX_X, PROXY_BOX_Y = prepare()


TOLERANCE = 5  # crop TL and DR tolerance radius


class Rt:  # runtime

    input_pth = None

    proxy_photo = None  # to prevent the image from being collected by the Python garbage collector


    ## 0: no rotation
    ## 1: 90 CCW
    ## 2: 180 deg
    ## 3: 90 CW
    rotate = 0


    do_crop = False
    crop_tl_x = None
    crop_tl_y = None
    crop_dr_x = None
    crop_dr_y = None
    crop_offset_x = None
    crop_offset_y = None
    crop_mode = None
    crop_proxy_w = None
    crop_proxy_h = None
    crop_proxy_x = None
    crop_proxy_y = None
    crop_real_w = None
    crop_real_h = None
    crop_w = None
    crop_h = None
    crop_x = None
    crop_y = None


    ## change these values to change the default value of the sliders.
    ## these values are the default by FFmpeg.
    contrast = 1
    brightness = 0
    saturation = 1
    gamma = 1
    gamma_r = 1
    gamma_g = 1
    gamma_b = 1
    vignette = 0
    colortemperature = 6500
    avgblur = 1  # 1 is the minimum value, based on ffmpeg error message

    ## the gate
    do_contrast = False
    do_brightness = False
    do_saturation = False
    do_gamma = False
    do_gamma_r = False
    do_gamma_g = False
    do_gamma_b = False
    do_vignette = False
    do_colortemperature = False
    do_avgblur = False

    ## 1: highest quality
    ## 31: lowest quality
    ## src: https://trac.ffmpeg.org/wiki/Encode/MPEG-4#:~:text=You%20can%20select%20a%20video,the%20lowest%20quality%2Fsmallest%20filesize.
    q_v = 1


def redraw_crop_grid():

    color = '#888'
    color_ctr = '#555'

    page.delete('crop_grid')

    ## because the center line darker, it must below the lighter
    y = Rt.crop_tl_y + (Rt.crop_dr_y-Rt.crop_tl_y)/2
    page.create_line(Rt.crop_tl_x, y, Rt.crop_dr_x, y, fill=color_ctr, tags='crop_grid')

    x = Rt.crop_tl_x + (Rt.crop_dr_x-Rt.crop_tl_x)/2
    page.create_line(x, Rt.crop_tl_y, x, Rt.crop_dr_y, fill=color_ctr, tags='crop_grid')

    page.create_rectangle(
        Rt.crop_tl_x, Rt.crop_tl_y, Rt.crop_dr_x, Rt.crop_dr_y,
        outline=color, tags='crop_grid'
    )
    page.create_oval(
        Rt.crop_tl_x-TOLERANCE, Rt.crop_tl_y-TOLERANCE,
        Rt.crop_tl_x+TOLERANCE, Rt.crop_tl_y+TOLERANCE,
        fill='#000', outline=color, tags='crop_grid'
    )
    page.create_oval(
        Rt.crop_dr_x-TOLERANCE, Rt.crop_dr_y-TOLERANCE,
        Rt.crop_dr_x+TOLERANCE, Rt.crop_dr_y+TOLERANCE,
        fill='#000', outline=color, tags='crop_grid'
    )

    y = Rt.crop_tl_y + (Rt.crop_dr_y-Rt.crop_tl_y)*(1/3)
    page.create_line(Rt.crop_tl_x, y, Rt.crop_dr_x, y, fill=color, tags='crop_grid')
    y = Rt.crop_tl_y + (Rt.crop_dr_y-Rt.crop_tl_y)*(2/3)
    page.create_line(Rt.crop_tl_x, y, Rt.crop_dr_x, y, fill=color, tags='crop_grid')

    x = Rt.crop_tl_x + (Rt.crop_dr_x-Rt.crop_tl_x)*(1/3)
    page.create_line(x, Rt.crop_tl_y, x, Rt.crop_dr_y, fill=color, tags='crop_grid')
    x = Rt.crop_tl_x + (Rt.crop_dr_x-Rt.crop_tl_x)*(2/3)
    page.create_line(x, Rt.crop_tl_y, x, Rt.crop_dr_y, fill=color, tags='crop_grid')


Label('software_title', 3, 3, f'{SOFTWARE_NAME}-v{SOFTWARE_VER}', 'Verdana 10', fg='#333')
core(
    page,
    Rt, SETTINGS_FFMPEG, SETTINGS_OPEN_DIR_PTH, SETTINGS_SAVE_DIR_PTH,
    PROXY_BOX_W, PROXY_BOX_H, PROXY_BOX_X, PROXY_BOX_Y,
    redraw_crop_grid
)


def left_mouse_press(e):
    Button.press_listener()
    Slider.press_listener()

    if Rt.do_crop:
        x, y = e.x, e.y
        if (abs(x-Rt.crop_tl_x) < TOLERANCE) and (abs(y-Rt.crop_tl_y) < TOLERANCE):  # resize using TL corner
            Rt.crop_mode = 'TL'
        elif (abs(x-Rt.crop_dr_x) < TOLERANCE) and (abs(y-Rt.crop_dr_y) < TOLERANCE):  # resize using DR corner
            Rt.crop_mode = 'DR'
        elif (Rt.crop_tl_x < x < Rt.crop_dr_x) and (Rt.crop_tl_y < y < Rt.crop_dr_y):  # to move the crop grid
            Rt.crop_mode = 'move'
            Rt.crop_offset_x = x - Rt.crop_tl_x
            Rt.crop_offset_y = y - Rt.crop_tl_y

root.bind('<ButtonPress-1>', left_mouse_press)

def left_mouse_hold(e):
    Slider.hold_listener()

    if Rt.do_crop:
        if Rt.crop_mode is not None:
            x, y = e.x, e.y

            ## minimum size for the crop grid
            min_w = 20
            min_h = 20

            if Rt.crop_mode == 'TL':
                ## these logics ensure the rectangle "crop" can't go beyond the image size and must has min area
                if (x < Rt.crop_dr_x - min_w) and (x >= Rt.crop_proxy_x):
                    Rt.crop_tl_x = x
                if (y < Rt.crop_dr_y - min_h) and (y >= Rt.crop_proxy_y):
                    Rt.crop_tl_y = y
            
            elif Rt.crop_mode == 'DR':
                if (x > Rt.crop_tl_x + min_w) and (x <= (Rt.crop_proxy_x + Rt.crop_proxy_w)):
                    Rt.crop_dr_x = x
                if (y > Rt.crop_tl_y + min_h) and (y <= (Rt.crop_proxy_y + Rt.crop_proxy_h)):
                    Rt.crop_dr_y = y
            
            elif Rt.crop_mode == 'move':
                new_tl_x = x - Rt.crop_offset_x
                new_tl_y = y - Rt.crop_offset_y
                new_dr_x = new_tl_x + (Rt.crop_dr_x - Rt.crop_tl_x)
                new_dr_y = new_tl_y + (Rt.crop_dr_y - Rt.crop_tl_y)
                if (new_tl_x >= Rt.crop_proxy_x) and (new_dr_x <= (Rt.crop_proxy_x + Rt.crop_proxy_w)):
                    Rt.crop_tl_x = new_tl_x
                    Rt.crop_dr_x = new_dr_x
                if (new_tl_y >= Rt.crop_proxy_y) and (new_dr_y <= (Rt.crop_proxy_y + Rt.crop_proxy_h)):
                    Rt.crop_tl_y = new_tl_y
                    Rt.crop_dr_y = new_dr_y

            redraw_crop_grid()
root.bind('<B1-Motion>', left_mouse_hold)

def left_mouse_release(e):
    Button.release_listener()
    Slider.release_listener()

    if Rt.do_crop:
        Rt.crop_mode = None
        Rt.crop_w = round( (Rt.crop_dr_x-Rt.crop_tl_x)*(Rt.crop_real_w/Rt.crop_proxy_w) )
        Rt.crop_h = round( (Rt.crop_dr_y-Rt.crop_tl_y)*(Rt.crop_real_h/Rt.crop_proxy_h) )
        Rt.crop_x = round( (Rt.crop_tl_x-Rt.crop_proxy_x)*(Rt.crop_real_w/Rt.crop_proxy_w) )
        Rt.crop_y = round( (Rt.crop_tl_y-Rt.crop_proxy_y)*(Rt.crop_real_h/Rt.crop_proxy_h) )
root.bind('<ButtonRelease-1>', left_mouse_release)

def background_fast():
    Button.hover_listener()
    Slider.hover_listener()
    root.after(50, background_fast)


def exit(e):
    printer('INFO: Exiting..')

    ## deleting intermediate files
    if os.path.isfile(PROXY_FILE_PTH):
        printer(f'INFO: Deleting {repr(PROXY_FILE_PTH)}...')
        os.remove(PROXY_FILE_PTH)

    root.destroy()
root.bind('<Escape>', exit)


def main():
    background_fast()
    root.mainloop()